rasa/shared/nlu/training_data/message.py

```diff
@@ -27,8 +28,12 @@
 if typing.TYPE_CHECKING:
     from rasa.shared.nlu.training_data.features import Features
 
+TEXT_TOKENS = TOKENS_NAMES.get(rasa.shared.nlu.constants.TEXT)
+
 
 class Message:
+    """Holds NLU training data."""
```

Comment:
I quickly jotted a description to stop the linter from complaining, I'm happy with any suggestions to improve this class definition.

Reply:
How about something like...
```
Container for data that describes one conversation state.

More precisely, the data describes a set of attributes that also can be used to describe at most 
one use and one action sub-state. For example, these attributes can be `TEXT`  and 
`INTENT` (describing a user utterance) or `ACTION_NAME` (describing a bot action). The container 
includes raw information (`self.data`) as well as features (`self.features`) for each such attribute. 
Moreover, the container can be tagged with a timestamp and can keep track about information 
on a specific  subset of attributes (`self.output_properties`).
```
using the [notion of state](https://www.notion.so/rasa/Events-States-1a6a181807d9421e9bd019deb6161f55#e04e8ce334754e82b31e7d4480a3796e) as it is at the moment and the term "features" for real (numeric unreadable) representations (which I think is sometimes also used for "attributes" - not that the terms "features" and "attributes" are really different -  but but we use them for very different things and should maybe use them like this (to avoid "featurized features" and instead have "features = featurization of (raw) attributes").

Reply:
I find "describes one conversation state" to be misleading, as this only captures the data of a single turn in a conversation, not the state of the conversation itself.

I also think a key piece of information about this class is it is used to pass data through the NLU pipeline and/or in end-to-end policies and that it's real purpose is to represent a user (or bot) _message_. We have other things which describe conversation state that are distinct from this as they are used in different places e.g. states from tracker featurization.

Reply:
Regarding conversation state vs turns: True, but you could also put information for both, user and action substate in there at the moment (and nothing would keep us from doing so). E.g. the logic in the TrackerFeaturizers even relies on the fact that the attributes of those substates don't overlap.

> We have other things which describe conversation state

Also true but at the moment it is also that sub-states are and must be of the same shape and form as what is passed to data as "message_data" and that substate names must correspond to the attribues (of a message).

Agree that I'm too focused here on describing what is there/could be done with it as opposed to motivating how it should be used though ^^. Feel free to rephrase.

How about...
```
Container for data that can be used to describe a conversation turn.

The turn is described by a set of attributes such as e.g. `TEXT`  and  `INTENT` when describing a user utterance
or e.g. `ACTION_NAME` for describing a bot action. The container includes raw information (`self.data`) as well as features (`self.features`) for each such attribute. 
Moreover, the message has a timestamp and can keep track about information on a specific  subset of attributes (`self.output_properties`).
```
@joejuzl ?

Reply:
Sounds good 👍